---
title: "Signac_mesoandcelltype"
output: html_document
date: "2024-08-15"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
# Load the required libraries
library(Signac)
library(Seurat)
library(GenomicRanges)
library(future)
library(BSgenome.Mmusculus.UCSC.mm10)
library(rtracklayer)
library(ggplot2)
library(dplyr)
library(data.table)

# Set parallel processing parameters
plan("multicore", workers = 20)
options(future.globals.maxSize = 70000 * 1024^2) # for 50 Gb RAM
```

```{r}
# Load input barcodes of organ mesothelia and cardiac cell types (cardiomyocytes and endothelial cells)
# Cell-type-specific clusters are identified through scATAC-seq analysis
lung <- read.delim('/mnt/sdb/qdang/meso_project/lung_meso.tsv', sep="\t", row.names = 1)
pancreas <- read.delim('/mnt/sdb/qdang/meso_project/pancreas_meso.tsv', sep="\t", row.names = 1)
heart <- read.delim('/mnt/sdb/qdang/atac/pycisTopic/E135_barcode.tsv', sep="\t", row.names = 1)
cm<- read.delim('/mnt/sdb/qdang/atac/sc_atac/E135/ventricular_cm.tsv', sep="\t", row.names = 1)
ec<- read.delim('/mnt/sdb/qdang/atac/sc_atac/E135/vascular_ec.tsv', sep="\t", row.names = 1)
```

```{r}
# create fragment objects based on fragment files
frags.lung <- CreateFragmentObject(
  path = "/mnt/sdb/qdang/meso_project/fragments/for_signac/lung_fragments.tsv.gz",
  cells = rownames(lung)
)

frags.pancreas <- CreateFragmentObject(
  path = "/mnt/sdb/qdang/meso_project/fragments/for_signac/pancreas_fragments.tsv.gz",
  cells = rownames(pancreas)
)

frags.epi <- CreateFragmentObject(
  path = "/mnt/sdb/qdang/meso_project/fragments/for_signac/E135_fragments.tsv.gz",
  cells = rownames(heart)
)

frags.ec <- CreateFragmentObject(
  path = "/mnt/sdb/qdang/meso_project/fragments/for_signac/vascularec_fragments.tsv.gz",
  cells = rownames(ec)
)

frags.cm <- CreateFragmentObject(
  path = "/mnt/sdb/qdang/meso_project/fragments/for_signac/ventricularcm_fragments.tsv.gz",
  cells = rownames(cm)
)
```

```{r}
# We have previously called consensus peaks of all five cell types using the peak extension and iterative peak filtering approach (Corces et al.)
peaks <- read.table(
  "consensus_regions.bed",row.names = NULL,
  col.names = c("chr", "start", "end","peakid","score","strand")
)
# Make a GRanges object 
gr <- makeGRangesFromDataFrame(peaks)
gr
```

```{r}
# Create a matrix of peaks x cell for each cell type
epi.counts <- FeatureMatrix(
  fragments = frags.epi,
  features = gr,
  cells = rownames(heart)
)

lung.counts <- FeatureMatrix(
  fragments = frags.lung,
  features = gr,
  cells = rownames(lung)
)

pancreas.counts <- FeatureMatrix(
  fragments = frags.pancreas,
  features = gr,
  cells = rownames(pancreas)
)

cm.counts <- FeatureMatrix(
  fragments = frags.cm,
  features = gr,
  cells = rownames(cm)
)

ec.counts <- FeatureMatrix(
  fragments = frags.ec,
  features = gr,
  cells = rownames(ec)
)
```

```{r}
# Create a Seurat object for each cell type
epi_assay <- CreateChromatinAssay(epi.counts, fragments = frags.epi)
epi_object <- CreateSeuratObject(epi_assay, assay = "ATAC", meta.data=heart)

lung_assay <- CreateChromatinAssay(lung.counts, fragments = frags.lung)
lung_object <- CreateSeuratObject(lung_assay, assay = "ATAC", meta.data=lung)

pancreas_assay <- CreateChromatinAssay(pancreas.counts, fragments = frags.pancreas)
pancreas_object <- CreateSeuratObject(pancreas_assay, assay = "ATAC", meta.data=pancreas)

cm_assay <- CreateChromatinAssay(cm.counts, fragments = frags.cm)
cm_object <- CreateSeuratObject(cm_assay, assay = "ATAC", meta.data=cm)

ec_assay <- CreateChromatinAssay(ec.counts, fragments = frags.ec)
ec_object <- CreateSeuratObject(ec_assay, assay = "ATAC", meta.data=ec)
```

```{r}
# add cell type annotation
epi_object$cell_type <- 'heart_mesothelium'
lung_object$cell_type <- 'lung_mesothelium'
pancreas_object$cell_type <- 'pancreas_mesothelium'
cm_object$cell_type <- 'ventricular_cm'
ec_object$cell_type <- 'vascular_ec'

# merge all datasets
combined <- merge(
  x = lung_object,
  y = list(pancreas_object, epi_object, cm_object, ec_object),
  add.cell.ids = c("lung", "pancreas", "epi", "cm", "ec")
)
combined[["ATAC"]]
```


```{r}
# Run TF-IDF normalisation, feature selection, dimensionality reduction using the SVD method, and perform UMAP
combined <- RunTFIDF(combined)
combined <- FindTopFeatures(combined, min.cutoff = 20)
combined <- RunSVD(combined)
combined <- RunUMAP(combined, dims = 2:50, reduction = 'lsi')
```


```{r}
# Add UCSC mm10 gene annotation to the dataset
mm10 <- rtracklayer::import('/mnt/sdb/qdang/reference/mouse/ucsc/gtf/mm10.ncbiRefSeq.gtf')
genome(mm10) <- "mm10"
seqlevelsStyle(mm10) <- "UCSC"
mm10$gene_biotype <- mm10$type
mm10$tx_id <- mm10$transcript_id
Annotation(combined) <- mm10
```

```{r}
# Add RNA modality data
counts <- readRDS("/mnt/sdb/qdang/meso_project/signac/combined_rna.rds")
```

```{r}
# Perform standard pre-processing steps. This include normalisation, feature selection, data scaling, PCA and UMAP
counts  <- NormalizeData(counts)
counts  <- FindVariableFeatures(counts)
counts  <- ScaleData(counts)
counts <- RunPCA(counts)
counts  <- RunUMAP(counts, dims = 1:30)
```

```{r}
# quantify gene activity
gene.activities <- GeneActivity(combined, features = VariableFeatures(counts), biotype = c("exon","CDS","transcript"))
```

```{r}
# add gene activities as a new assay
combined[["ACTIVITY"]] <- CreateAssayObject(counts = gene.activities)

# normalize gene activities
DefaultAssay(combined) <- "ACTIVITY"
combined <- NormalizeData(combined)
combined <- ScaleData(combined, features = rownames(combined))
```


```{r}
# Identify anchors
transfer.anchors <- FindTransferAnchors(reference = counts, query = combined, features = VariableFeatures(object = counts),
    reference.assay = "originalexp", query.assay = "ACTIVITY", reduction = "cca")
```


```{r}
# Process the enhancer-promoter (E-P) linked predicted by the Activity-By-Contact method for incorporation into the data
DefaultAssay(combined) <- "ATAC"
bed <- as.data.frame(read.table("/mnt/sdb/qdang/atac/abc_enhancer/652024_revisedrun/predict/E135/links_for_signac.bed",header = FALSE, sep="\t",stringsAsFactors=FALSE, quote=""))
```

```{r}
# Process the E-P file into a format compatible with Signac
bed2 <- subset(bed[bed$V2 < bed$V3, ])
bed3 <- subset(bed[bed$V2 > bed$V3, ])
bed3 <- bed3 %>% relocate(V2, .after=V3)
dataList <- list(bed2, bed3)
bed4 <- rbindlist(dataList, use.names=FALSE)
bed$V6 <- NULL
names(bed4)[names(bed4) == 'V5'] <- 'score'
```

```{r}
# Convert E-P links to a GRanges dataframe
link <- makeGRangesFromDataFrame(bed4, 
        seqnames.field="V1",
        start.field="V2",
        end.field="V3",
        keep.extra.columns=TRUE)

# Incorporate the links into the Signad dataframe
Links(combined) <- link
Links(combined)
```

```{r}
Idents(object = combined) <- "cell_type"
```

```{r}
# Add .bigwig files for visualition. Here, we are visualising H3K27ac and H3K4me3 histone modification data of MEC1 epicardial cells. 
listbigwig <- list("/mnt/sdb/qdang/cut_run/H3K27ac_EPI_merged_CPM.bw",
               "/mnt/sdb/qdang/cut_run/H3K4me3_EPI_merged_CPM.bw")
names(listbigwig) <- c("H3K27ac", "H3K4me3")
```

```{r}
cov_plot <- CoveragePlot(
  object = combined,
  region = "chr11-35034398-35163411",
  annotation = FALSE,
  group.by='cell_type',
  peaks = FALSE,
    bigwig=listbigwig,
  bigwig.scale = 'separate',
  links=FALSE
)
cov_plot
```


```{r}
cov_plot <- CoveragePlot(
  object = combined,
  region = "chr11-96277840-96285902",
  annotation = FALSE,
  group.by='cell_type',
  peaks = FALSE,
  links=FALSE
)
cov_plot
#chr8-80056416-80060688: Hhip promoter regulated by Gli1
```


```{r}
gene_plot <- AnnotationPlot(
  object = combined,
  region = "chr11-35034398-35163411"
)
gene_plot
```


```{r}
link_plot <- LinkPlot(
  object = combined,
  region = "chr11-35034398-35163411"
)
link_plot
```


```{r}
gene <- CombineTracks(
  plotlist = list(cov_plot, gene_plot, link_plot),
  heights = c(45, 7, 7),
  widths = c(10, 1)
)
gene
```

```{r}
# Below, we analyse pancreas scATAC-seq data spanning embryonic development
pancreas_e145 <- read.delim('/mnt/sdb/qdang/meso_project/sc_atac/e145_pancreas_meso.tsv', sep="\t", row.names = 1)
pancreas_e145_2 <- read.delim('/mnt/sdb/qdang/meso_project/sc_atac/e145_pancreas_2_meso.tsv', sep="\t", row.names = 1)
pancreas_e135 <- read.delim('/mnt/sdb/qdang/meso_project/pancreas_meso.tsv', sep="\t", row.names = 1)
pancreas_e175 <- read.delim('/mnt/sdb/qdang/meso_project/sc_atac/e175_pancreas_meso.tsv', sep="\t", row.names = 1)
pancreas_e175_2 <- read.delim('/mnt/sdb/qdang/meso_project/sc_atac/e175_pancreas_2_lane_1_meso.tsv', sep="\t", row.names = 1)
```

```{r}
# create fragment objects

frags.e135 <- CreateFragmentObject(
  path = "/mnt/sdb/qdang/meso_project/fragments/for_signac/pancreas_fragments.tsv.gz",
  cells = rownames(pancreas_e135)
)

frags.e145 <- CreateFragmentObject(
  path = "/mnt/sdb/qdang/meso_project/sc_atac/mouse_e145_pancreas_fragments.tsv.gz",
  cells = rownames(pancreas_e145)
)

frags.e145.2 <- CreateFragmentObject(
  path = "/mnt/sdb/qdang/meso_project/sc_atac/mouse_e145_pancreas_2_fragments.tsv.gz",
  cells = rownames(pancreas_e145_2)
)

frags.e175 <- CreateFragmentObject(
  path = "/mnt/sdb/qdang/meso_project/sc_atac/mouse_e175_pancreas_fragments.tsv.gz",
  cells = rownames(pancreas_e175)
)

frags.e175.2 <- CreateFragmentObject(
  path = "/mnt/sdb/qdang/meso_project/sc_atac/mouse_e175_pancreas_2_fragments.tsv.gz",
  cells = rownames(pancreas_e175_2)
)
```

```{r}
peaks <- read.table(
  file = "/mnt/sdb/qdang/meso_project/sc_atac/outs/consensus_peak_calling/consensus_regions.bed",row.names = NULL,
  col.names = c("chr", "start", "end","peakid","score","strand")
)

gr <- makeGRangesFromDataFrame(peaks)
gr
```

```{r}
e145.counts <- FeatureMatrix(
  fragments = frags.e145,
  features = gr,
  cells = rownames(pancreas_e145)
)

e145.2.counts <- FeatureMatrix(
  fragments = frags.e145.2,
  features = gr,
  cells = rownames(pancreas_e145_2)
)

e135.counts <- FeatureMatrix(
  fragments = frags.e135,
  features = gr,
  cells = rownames(pancreas_e135)
)

e175.counts <- FeatureMatrix(
  fragments = frags.e175,
  features = gr,
  cells = rownames(pancreas_e175)
)

e175.2.counts <- FeatureMatrix(
  fragments = frags.e175.2,
  features = gr,
  cells = rownames(pancreas_e175_2)
)
```

```{r}
e135_assay <- CreateChromatinAssay(e135.counts, fragments = frags.e135)
e135_object <- CreateSeuratObject(e135_assay, assay = "ATAC", meta.data=pancreas_e135)

e145_assay <- CreateChromatinAssay(e145.counts, fragments = frags.e145)
e145_object <- CreateSeuratObject(e145_assay, assay = "ATAC", meta.data=pancreas_e145)

e145_2_assay <- CreateChromatinAssay(e145.2.counts, fragments = frags.e145.2)
e145_2_object <- CreateSeuratObject(e145_2_assay, assay = "ATAC", meta.data=pancreas_e145_2)

e175_assay <- CreateChromatinAssay(e175.counts, fragments = frags.e175)
e175_object <- CreateSeuratObject(e175_assay, assay = "ATAC", meta.data=pancreas_e175)

e175_2_assay <- CreateChromatinAssay(e175.2.counts, fragments = frags.e175.2)
e175_2_object <- CreateSeuratObject(e175_2_assay, assay = "ATAC", meta.data=pancreas_e175_2)
```


```{r}
# add information to identify dataset of origin
e145_object$sample <- 'E14.5'
e145_2_object$sample <- 'E14.5 rep 2'
e175_object$sample <- 'E17.5'
e175_2_object$sample <- 'E17.5 rep 2'
e135_object$time <- 'E13.5'
e145_object$time <- 'E14.5'
e145_2_object$time <- 'E14.5'
e175_object$time <- 'E17.5'
e175_2_object$time <- 'E17.5'


# merge all datasets, adding a cell ID to make sure cell names are unique
combined <- merge(
  x = e135_object,
  y = list(e145_object, e145_2_object, e175_object, e175_2_object),
  add.cell.ids = c("E135", "E145", "E145_2", "E175", "E175_2")
)
combined[["ATAC"]]
```

```{r}
combined <- RunTFIDF(combined)
combined <- FindTopFeatures(combined, min.cutoff = 20)
combined <- RunSVD(combined)
combined <- RunUMAP(combined, dims = 2:50, reduction = 'lsi')
```

```{r}
DimPlot(combined, group.by = 'sample', pt.size = 0.1)
```

```{r}
mm10 <- rtracklayer::import('/mnt/sdb/qdang/reference/mouse/ucsc/gtf/mm10.ncbiRefSeq.gtf')
genome(mm10) <- "mm10"
seqlevelsStyle(mm10) <- "UCSC"
mm10$gene_biotype <- mm10$type
mm10$tx_id <- mm10$transcript_id
Annotation(combined) <- mm10
```


```{r}
# Analyse embryonic lung ATAC-seq samples
lung_e135 <- read.delim('/mnt/sdb/qdang/meso_project/lung_meso.tsv', sep="\t", row.names = 1)
lung_p3 <- read.delim('/mnt/sdb/qdang/meso_project/sc_atac/lung/p3_lung_meso.tsv', sep="\t", row.names = 1)
```

```{r}
# create fragment objects

frags.e135 <- CreateFragmentObject(
  path = "/mnt/sdb/qdang/meso_project/fragments/for_signac/lung_fragments.tsv.gz",
  cells = rownames(lung_e135)
)


frags.p3 <- CreateFragmentObject(
  path = "/mnt/sdb/qdang/meso_project/sc_atac/lung/PMID_33707239/GSM4504968_P3_fragments.tsv.gz",
  cells = rownames(lung_p3)
)
```

```{r}
peaks <- read.table(
  file = "/mnt/sdb/qdang/meso_project/sc_atac/lung/outs/consensus_peak_calling/consensus_regions.bed",row.names = NULL,
  col.names = c("chr", "start", "end","peakid","score","strand")
)

gr <- makeGRangesFromDataFrame(peaks)
gr
```

```{r}
e135.counts <- FeatureMatrix(
  fragments = frags.e135,
  features = gr,
  cells = rownames(lung_e135)
)

p3.counts <- FeatureMatrix(
  fragments = frags.p3,
  features = gr,
  cells = rownames(lung_p3)
)
```

```{r}
e135_assay <- CreateChromatinAssay(e135.counts, fragments = frags.e135)
e135_object <- CreateSeuratObject(e135_assay, assay = "ATAC", meta.data=lung_e135)

p3_assay <- CreateChromatinAssay(p3.counts, fragments = frags.p3)
p3_object <- CreateSeuratObject(p3_assay, assay = "ATAC", meta.data=lung_p3)
```


```{r}
# add information to identify dataset of origin
e135_object$sample <- 'E13.5'
p3_object$sample <- 'P3'


# merge all datasets, adding a cell ID to make sure cell names are unique
combined <- merge(
  x = e135_object,
  y = p3_object,
  add.cell.ids = c("E135", "P3")
)
combined[["ATAC"]]
```

```{r}
combined <- RunTFIDF(combined)
combined <- FindTopFeatures(combined, min.cutoff = 20)
combined <- RunSVD(combined)
combined <- RunUMAP(combined, dims = 2:50, reduction = 'lsi')
```

```{r}
DimPlot(combined, group.by = 'sample', pt.size = 0.1)
```

```{r}
mm10 <- rtracklayer::import('/mnt/sdb/qdang/reference/mouse/ucsc/gtf/mm10.ncbiRefSeq.gtf')
genome(mm10) <- "mm10"
seqlevelsStyle(mm10) <- "UCSC"
mm10$gene_biotype <- mm10$type
mm10$tx_id <- mm10$transcript_id
Annotation(combined) <- mm10
```
